### PR TITLE
Epic/icml

### PR DESCRIPTION
**Pull request by giovannic**
## Summary                                                      
  ICML epic branch: adds end-to-end TFMPE estimators (direct, bottom-up, posterior factorisation), L-C2ST metrics, transformer improvements (linear attention, learned embeddings, mixed precision, cuDNN +      
  checkpointing), and all the tests.                                                                                                                                                                  
   
  ## Changes                                                                                                                                                                                                     
  ### New Features
  - **Estimators**: new `tfmpe/estimators/` module with `tfmpe.py`, `ode.py`, `proposals.py`, and a `training/` package covering `directly`, `bottom_up`, `posterior_factorisation`, and a shared `loss` module.
  - **Metrics**: `tfmpe/metrics/lc2st.py` plus a new `nn/classifier.py` for L-C2ST evaluation.                                                                                                                   
  - **NN**: `nn/training.py` training utilities; `transformer/linear_attention.py`; learned position/group embeddings and group-id support in tokens; alternative ODE solvers; mixed precision; cuDNN attention  
  with gradient checkpointing.                                                                                                                                                                                   
  - **Preprocessing**: `pad_to_even` for tokens and combine; group-id plumbing.                                                                                                                                  
  - **Sampling**: scan-based batched posterior sampling with parameterisable batch size; stick-breaking for small sample sizes in PF.                                                                            
  - **Tests**:  `test/test_estimators/` for flow matching specific tests (ode, loss, e2e...), `test/test_metrics/` for l2st specific tests, `test/test_nn/` for more transformer tests.                                             
  - **Benchmarks/ablations**: PF ablation, transformer benchmarks, dtype tests.                                                                                                                                  
                                                                                                                                                                                                                 
  ### Breaking Changes                                                                                                                                                                                           
  - `tfmpe/estimators/__init__.py` exports reshaped around the new estimator API.                                                                                                                                
  - Token/embedding interfaces extended with `group_ids` and even-padding; downstream callers must pass the new fields.                                                                                          
  - Removed `tfmpe/preprocessing/utils.py` helpers                                                                                                                              
                                                                                                                                                                                                                 
  ### Fixes                                                                                                                                                                                                      

None                                                                                                                                     
                                                                                                                                                                                                                 
  ---                                                                                                                                                                                                            
                                                                                                                                                                                                                 
  ## Reviewer Checklist                                                                                                                                                                                          
   
  ### All PRs                                                                                                                                                                                                    
  - [ ] **Correctness**: Logic is sound and implementation is correct
  - [ ] **Testing**: Appropriate tests added/updated                                                                                                                                                             
  - [ ] **Benchmarking**: Training/sampling speed impact assessed if relevant                                                                                                                                    
  - [ ] **Documentation**: Code comments and docs updated where needed                                                                                                                                           
                                                                                                                                                                                                                 
  ### For AI-Generated PRs                                                                                                                                                                                       
  - [ ] **Entropy**: No unnecessary code duplication, verbose implementations, or misplaced imports                                                                                                              
  - [ ] **Test Quality**: Tests genuinely validate intended behavior (check `.claude/spec/{feature}/tasks.md` for requirements)

### CODE REVIEW

File: tfmpe/estimators/training/posterior_factorisation.py
Comment (absternator) [inline tfmpe/estimators/training/posterior_factorisation.py:1]: This isn't used anywhere. (except tests).  would it be used in the future?

File: test/test_metrics/test_lc2st_e2e.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    # Bad: incorrect sampling (just adds noise to truth, ignores observation weighting)
+    key_bad_obs, key = jr.split(key)
+    theta_post_bad = theta_truth + jr.normal(key_bad_obs, (train_size, dim)) * sigma + 1
+
+    r_good = run_lc2st(ev_key, d_cal_good, observation, theta_post_good)
+    r_bad = run_lc2st(ev_key, d_cal_bad, observation, theta_post_bad)
+
+    assert r_good.main_stat < r_bad.main_stat
+    # assert r_good.main_stat < r_good.critical_value(0.95)
```
Comment (absternator) [inline test/test_metrics/test_lc2st_e2e.py]: commented out code

File: tfmpe/nn/transformer/config.py [hunk 2]
```diff
@@ -48,8 +50,14 @@ class TransformerConfig:
     label_dim: int = 32
     index_out_dim: int = 64
     pos_dim: int = 8
-    dropout: float = 0.1
+    max_positions: int = 128
+    group_dim: int = 8
+    max_groups: int = 128
```
Comment (absternator) [inline tfmpe/nn/transformer/config.py]: should probably force people to pass this in, otherwise they will have issues if their groups > 128... like the issue we discovered when I ran with 1000 sites

File: tfmpe/estimators/training/bottom_up.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    epsilon: float = 1e-3,
+    sample_batch_size: int = 1000
+) -> Tuple[TFMPE, List[Tuple[Array, Array, Array, Array]]]:
+    """Multi-round bottom-up training algorithm.
+
+    Each round alternates between local likelihood training
+    (n=1 local groups) and global posterior training
+    (n=n_groups local groups).
+
+    Currently only supports n_rounds=1. Each round makes two
```
Comment (absternator) [inline tfmpe/estimators/training/bottom_up.py]: when would we have multiple rounds? and if it only supports one currently should we remove the loop etc to avoid confusion? unless multiround is going to happen soon?

File: tfmpe/estimators/training/bottom_up.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    n_val_samples: int,
+    opt: nnx.Optimizer | List[nnx.Optimizer],
+    n_iter_per_round: int,
+    batch_size: int,
+    rng: PRNGKeyArray,
+    labeller: Labeller,
+    prior_log_prob: Callable[[PyTree], float],
+    prob_transform: Optional[Callable] = None,
+    obs_f_in: Optional[Dict] = None,
+    f_in_fn: Optional[Callable]=None,
```
Comment (absternator) [inline tfmpe/estimators/training/bottom_up.py:40]: what is this used for? `f_in_fn`

File: tfmpe/estimators/training/bottom_up.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    (n=1 local groups) and global posterior training
+    (n=n_groups local groups).
+
+    Currently only supports n_rounds=1. Each round makes two
+    training calls:
+    1. Train p(y|theta) with n=1 local parameters
+    2. Train p(theta,z|y) with n=n_groups local parameters
+
+    Parameters
+    ----------
```
Comment (absternator) [inline tfmpe/estimators/training/bottom_up.py:59]: doc bloc needs updating

File: tfmpe/estimators/training/bottom_up.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        val_tokens = Tokens.from_pytree(
+            {**val_y, **val_theta},
+            condition=list(val_theta.keys()),
+            labeller=labeller,
+            sample_ndims=1,
+            functional_inputs=val_f_in
+        )
+
+        # First fit: p(y|theta) with n=1
+        tfmpe_local.train()
```
Comment (absternator) [inline tfmpe/estimators/training/bottom_up.py:266]: fitting does train and validation right. so makes more sense to set `   tfmpe_local.train()` right befire training loop and then `   tfmpe_local.eval()` before validation loop.

File: tfmpe/estimators/training/directly.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+from ..tfmpe import TFMPE
+from ...preprocessing.tokens import Tokens
+from ...preprocessing.utils import Labeller
+from ...nn.training import fit_nn
+
+from .loss import cfm_loss as _cfm_loss
+
+
+def fit_directly(
```
Comment (absternator) [inline tfmpe/estimators/training/directly.py:17]: just to duble check the main differnce here is we simulate y and not fit (y | simulated theta)

File: tfmpe/estimators/proposals.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
+from typing import Callable, Optional, Dict, Tuple
+from jaxtyping import Array, PyTree
+
+from jax import tree, numpy as jnp, random as jr
+
+from .tfmpe import TFMPE
+from ..preprocessing.utils import Labeller
+from ..preprocessing.tokens import Tokens
+
+def truncated_proposal_rejection(
```
Comment (absternator) [inline tfmpe/estimators/proposals.py:10]: what is this used for ? cant see any reference?

File: tfmpe/estimators/tfmpe.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+        ----------
+        vf_network : TokenisedVectorField
+            Vector field network f(context, params, t) -> Array.
+            Must be an nnx.Module so its state is captured and
+            RNG streams are properly managed during training.
+        base_dist : BaseDistribution
+            Base distribution with sample(shape) and
+            log_prob(x) methods. Example: NormalDistribution(rngs)
+        solver : diffrax solver, optional
+            ODE solver instance. Default: Heun()
```
Comment (absternator) [inline tfmpe/estimators/tfmpe.py]: doc update

File: tfmpe/metrics/lc2st.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        patience=100
+    )
+
+    # Train null classifiers with permuted labels
+    n_null = null_classifier.n
+    rng_key, null_key = jr.split(rng_key)
+
+    # Create 3D input for null classifiers: (batch_size, n_classifiers, feature_dim)
+    u_null_base = jnp.concatenate([u_joint, u_posterior], axis=0)  # (2*N_cal, feature_dim)
+    u_null = jnp.broadcast_to(
```
Comment (absternator) [inline tfmpe/metrics/lc2st.py:134]:
```suggestion
    u_null = jnp.broadcast_to(
        u_null_base[:, None, :],
        (2*N_cal, n_null, u_null_base.shape[-1])) # (2*N_cal, n_classifiers, feature_dim)
```
as discussed to stop materialization

File: tfmpe/nn/transformer/config.py (hunk 2 above)
Comment (absternator) [reply tfmpe/nn/transformer/config.py]: probably same with max positions fi needed

File: tfmpe/nn/transformer/embedding.py [hunk 11]
```diff
@@ -86,10 +88,14 @@ def __init__(
         n_labels: int,
         label_dim: int,
         pos_dim: int,
+        max_positions: int,
         latent_dim: int,
         rngs: nnx.Rngs,
         f_in_in_dim: int = 0,
         f_in_out_dim: int = 0,
+        group_dim: int = 0,
+        max_groups: int = 128,
```
Comment (absternator) [inline tfmpe/nn/transformer/embedding.py:97]: same as before should probs pass in

File: tfmpe/nn/transformer/embedding.py [hunk 12]
```diff
@@ -102,35 +108,64 @@ def __init__(
...
+        self.group_dim = group_dim
+        if group_dim > 0:
+            self.group_emb = nnx.Embed(
+                max_groups,
+                features=group_dim,
+                dtype=ops_dtype,
+                rngs=rngs,
+            )
 
         if f_in_in_dim > 0:
```
Comment (absternator) [inline tfmpe/nn/transformer/embedding.py:155]: what is `f_in_in` again?

File: tfmpe/nn/transformer/linear_attention.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
+"""Linear attention implementation."""
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array
+from flax.typing import Dtype, PrecisionLike
+from flax.nnx.module import Module
+
+
+def linear_attention(
```
Comment (absternator) [inline tfmpe/nn/transformer/linear_attention.py:10]: very clean implementation !!!

File: tfmpe/nn/classifier.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        @nnx.vmap(in_axes=(0, 1), out_axes=1)
+        @nnx.vmap(in_axes=(None, 1), out_axes=1)
+        def call_classifier_on_slice(cls, u_slice):
+            return cls(u_slice)
+
+        return call_classifier_on_slice(self.classifiers, u)
+
+Classifier = MultiBinaryMLPClassifier | BinaryMLPClassifier | MultiFoldBinaryMLPClassifier
+
+def ce_loss(
```
Comment (absternator) [inline tfmpe/nn/classifier.py:137]: should this be in `loss.py`

File: tfmpe/nn/training.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
+from jax import numpy as jnp, random as jr, tree
+from flax import nnx
+from jaxtyping import PyTree, Array
+from typing import Tuple, Callable, List, Optional, Any, TypeVar
+from tqdm import tqdm
+
+M = TypeVar('M', bound=nnx.Module)
+
+def fit_nn(
```
Comment (absternator) [inline tfmpe/nn/training.py:9]: Q: how come the `fit_fast` is removed? was it not working or too much memory consumption?

File: tfmpe/nn/training.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+            batch_losses.append(batch_loss)
+
+        # Average training loss across batches
+        train_loss = jnp.mean(jnp.stack(batch_losses))
+        train_losses_list.append(train_loss)
+
+        # Compute validation loss
+        if val is not None:
+            epoch_rng, val_key = jr.split(epoch_rng)
+
```
Comment (absternator) [inline tfmpe/nn/training.py:145]: `model.eval()` needed?

File: tfmpe/nn/training.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+            val_loss = compute_val_loss(model, val, val_key)
+            val_losses_list.append(val_loss)
+
+            # Update progress bar with losses
+            pbar.set_postfix(train_loss=f"{float(train_loss):.4f}", val_loss=f"{float(val_loss):.4f}")
+
+            # Early stopping check
+            if early_stopping_enabled:
+                if best_val_loss - float(val_loss) > delta:
```
Comment (absternator) [inline tfmpe/nn/training.py]: can extract `float(val_loss)` so dont have to double host sync

File: tfmpe/preprocessing/tokens.py [hunk 18]
```diff
@@ -279,19 +306,59 @@ def _from_pytree_impl(
             broadcast_shape
         )
 
+        padding_mask = None
+
+        # Pad to even sequence length for cuDNN flash attention
+        if pad_to_even and total_tokens % 2 == 1:
```
Comment (absternator) [inline tfmpe/preprocessing/tokens.py:312]: as discussed we should probs make a masking/padding class to handle this. just leaving note for future. not this pr

File: pyproject.toml
Comment (absternator) [inline pyproject.toml:1]: i had problems running with higher depencies of jax and flax should tie down versions. below is what i found worked
```
 "jax[cuda12]==0.9.2",
    "flax==0.12.6",

**Review by absternator — COMMENTED** [review]
Nice work!! works well as ive been running for the past weeks! just left some comments/questions/

File: tfmpe/nn/transformer/config.py (hunk 2 above)
Comment (giovannic) [reply tfmpe/nn/transformer/config.py]: The thing is at the moment, model configuration is separate from tokenisation and I think that's good. Perhaps we could put a warning in `fit_nn`? We could inspect the model and the tokens and print a warning if they don't line up?

File: tfmpe/estimators/training/bottom_up.py (hunk 3 above)
Comment (giovannic) [reply tfmpe/estimators/training/bottom_up.py]: Oh it does support multiple rounds (sequential ablation) the docs should be updated.

File: tfmpe/estimators/training/bottom_up.py (hunk 4 above)
Comment (giovannic) [reply tfmpe/estimators/training/bottom_up.py:40]: This is used for functional observations, the SEIR part of the paper: https://github.com/giovannic/sbijax/blob/feat/hierarchical_A/examples/seir.py#L118

File: tfmpe/estimators/training/directly.py (hunk 7 above)
Comment (giovannic) [reply tfmpe/estimators/training/directly.py:17]: Yes exactly, to compare to the errors introduced by approximating \hat{y} \sim p(y | theta) in bottom up

File: tfmpe/estimators/proposals.py (hunk 8 above)
Comment (giovannic) [reply tfmpe/estimators/proposals.py:10]: Oh, the user can decide to use this instead of sir sampling when calling bottom_up. This was a method I did not include because it distracted from the main point of the paper and requires a lot of work to make good (probably Tim's work). I've not removed it because it's interesting research haha

File: tfmpe/metrics/lc2st.py (hunk 10 above)
Comment (giovannic) [reply tfmpe/metrics/lc2st.py:134]: would take this but it looks like the indentation is off?

File: tfmpe/nn/training.py (hunk 15 above)
Comment (giovannic) [reply tfmpe/nn/training.py:9]: Early stopping became essential to good performance. Fit_fast was too inflexible and wasn't much quicker in my experiments. My guess is that there's not much time to be saved in compiling the loop since the bottleneck is in the actual backprop and update instead of dispatch. And I didn't want to maintain 2 training loops.

File: tfmpe/nn/classifier.py (hunk 14 above)
Comment (giovannic) [reply tfmpe/nn/classifier.py:137]: Can we keep this in here? tfmpe/estimator/training/loss.py is for tfmpe/estimator/training/*.py. This loss is for direct use with tfmpe/nn/training.py

File: tfmpe/metrics/lc2st.py (hunk 10 above)
Comment (absternator) [reply tfmpe/metrics/lc2st.py:134]: oh yeah haha yeah ive fixed. but yeah just replace the u_null assigment

**Review by absternator — APPROVED** [review]
Thanks lgtm! just a couple requests but happy to approve after that

File: tfmpe/estimators/training/posterior_factorisation.py
Comment (giovannic) [reply tfmpe/estimators/training/posterior_factorisation.py:1]: This is used as the "PF" ablation in sbibm!
